### ISSUE TITLE
error on call to consolidate

### ISSUE THREAD

**Issue by bkmartinjr** [TileDB-Inc/TileDB-Py#230]
With the most recent release, I am getting an error on consolidate.  Something odd seems to be happening on the second dimension (based upon the error message).

```
Traceback (most recent call last):
  File "./bug.py", line 38, in <module>
    main()
  File "./bug.py", line 34, in main
    tiledb.consolidate(name)
  File "tiledb/libtiledb.pyx", line 4420, in tiledb.libtiledb.consolidate
  File "tiledb/libtiledb.pyx", line 387, in tiledb.libtiledb._raise_ctx_err
  File "tiledb/libtiledb.pyx", line 372, in tiledb.libtiledb._raise_tiledb_error
tiledb.libtiledb.TileDBError: [TileDB::Query] Error: Subarray out of bounds. subarray: [0, 27999, 0, 20699] domain: [0, 27999, 0, 20645]
```

To reproduce:
```
import tiledb
import numpy as np


def make_array(name, shape):

    filters = tiledb.FilterList([
        tiledb.ZstdFilter(),
    ])
    attrs = [
        tiledb.Attr(dtype=np.float32, filters=filters)
    ]
    domain = tiledb.Domain(tiledb.Dim(name="obs", domain=(0, shape[0] - 1), tile=min(shape[0], 200), dtype=np.uint32),
                           tiledb.Dim(name="var", domain=(0, shape[1] - 1), tile=min(shape[1], 100), dtype=np.uint32))

    schema = tiledb.ArraySchema(domain=domain, sparse=False, attrs=attrs,
                                cell_order='row-major', tile_order='row-major')
    tiledb.DenseArray.create(name, schema)


def main():

    shape = (28000, 20646)
    name = "X"
    make_array(name, shape)

    stride = int(np.power(10, np.around(np.log10(1e8 / shape[1]))))
    with tiledb.DenseArray(name, mode='w') as X:
        for row in range(0, shape[0], stride):
            lim = min(row+stride, shape[0])
            print(row, lim)
            X[row:lim, :] = np.random.rand(lim-row, shape[1])

        tiledb.consolidate(name)


if __name__ == '__main__':
    main()

```

**Comment by stavrospapadopoulos** [comment]
@bkmartinjr we identified the issue, thanks for raising it. It is an internal issue with the way dense tiles must be read in their entirety to be written to the new consolidated fragment. Your domain is not divisible by the tile extent and this is causing an issue. We'll fix this shortly.

In the meantime, could you please expand your column domain to be divisible by the column tile extent? This will fix the problem for now. After my path, you will again be able to define arbitrary domains (even non-divisible by the tile extents).

**Comment by stavrospapadopoulos** [comment]
@bkmartinjr this is fixed by https://github.com/TileDB-Inc/TileDB/pull/1442. @ihnorton will incorporate the patch in TileDB-Py. Thanks again.

**Comment by ihnorton** [comment]
Hi @bkmartinjr, TileDB-Py 0.5.1 packages, including the fix mentioned by Stavros, are available from pip. I successfully tested your script in a clean install on linux. Conda packages should be available in ~1 hr.

**Comment by bkmartinjr** [comment]
Thank you for the fast fix.  I have verified that it works on my end as well.

**Comment by bkmartinjr** [comment]
I spoke too soon - there is still a bug related to the same core issue.  This is the same test case, followed by code that accesses the array.  It fails:

```
Traceback (most recent call last):
  File "bug.py", line 45, in <module>
    main()
  File "bug.py", line 40, in main
    d = A.multi_index[rows, :]['']
  File "/mnt/c/Users/bruce/projects/tiledb-fun/.venv/lib/python3.6/site-packages/tiledb/multirange_indexing.py", line 120, in __getitem__
    arr.shape = result_shape
ValueError: cannot reshape array of size 516150016 into shape (25000,20646)
```

Extended test case:
```
import tiledb
import numpy as np
import random


def make_array(name, shape):

    filters = tiledb.FilterList([
        tiledb.ZstdFilter(),
    ])
    attrs = [
        tiledb.Attr(dtype=np.float32, filters=filters)
    ]
    domain = tiledb.Domain(tiledb.Dim(name="obs", domain=(0, shape[0] - 1), tile=min(shape[0], 200), dtype=np.uint32),
                           tiledb.Dim(name="var", domain=(0, shape[1] - 1), tile=min(shape[1], 100), dtype=np.uint32))

    schema = tiledb.ArraySchema(domain=domain, sparse=False, attrs=attrs,
                                cell_order='row-major', tile_order='row-major')
    tiledb.DenseArray.create(name, schema)


def main():

    shape = (28000, 20646)
    name = "X"
    make_array(name, shape)

    stride = int(np.power(10, np.around(np.log10(1e8 / shape[1]))))
    with tiledb.DenseArray(name, mode='w') as X:
        for row in range(0, shape[0], stride):
            lim = min(row+stride, shape[0])
            print(row, lim)
            X[row:lim, :] = np.random.rand(lim-row, shape[1])

        tiledb.consolidate(name)

    # Additional code
    rows = random.sample(range(0, shape[0]), k=25_000)
    rows.sort()
    with tiledb.DenseArray(name) as A:
        d = A.multi_index[rows, :]['']
        d.mean(axis=0)


if __name__ == '__main__':
    main()
```

**Comment by bkmartinjr** [comment]
This bug seems to have re-appeared in the latest release (0.5.8).    The dataset it is occurring on is quite huge.  If you want a test case, LMK and I'll try to work up one.

```
  File "../cxgtool.py", line 364, in save_X
    tiledb.consolidate(X_name, ctx=ctx)
  File "tiledb/libtiledb.pyx", line 4678, in tiledb.libtiledb.consolidate
  File "tiledb/libtiledb.pyx", line 413, in tiledb.libtiledb._raise_ctx_err
  File "tiledb/libtiledb.pyx", line 398, in tiledb.libtiledb._raise_tiledb_error
tiledb.libtiledb.TileDBError: [TileDB::Query] Error: Subarray out of bounds. subarray: [0, 1163899, 0, 33299] domain: [0, 1163850, 0, 33294]
```

**Comment by ihnorton** [comment]
@bkmartinjr, just to double-check: did the code work with 0.5.6, or was that blocked by #290? (I checked the test-case in #231 -- still works)

I'm happy to debug real code/data with a debug build, but I realize there might be access limitations on the data. Otherwise, if you can repro easily, that would be great. Or share the schema and what the write pattern looks like and I'll try to repro.

**Comment by bkmartinjr** [comment]
I confirmed that the original issue appears to be fixed in 0.5.8 and 0.5.6, and this new issue (same symptoms) occurs in both.

So I'll work on a test case and will add here.

**Comment by ihnorton** [comment]
Is the array sparse?

**Comment by bkmartinjr** [comment]
Dense

**Comment by bkmartinjr** [comment]
Dataset is very large, so it will take me a few to get a test case.  But here is a dump() on the array.
```
>>> A.dump()
- Array type: dense
- Cell order: row-major
- Tile order: col-major
- Capacity: 10000
- Coordinates compressor: ZSTD
- Coordinates compression level: -1

=== Domain ===
- Dimensions type: UINT32

### Dimension ###
- Name: obs
- Domain: [0,1163850]
- Tile extent: 50

### Dimension ###
- Name: var
- Domain: [0,33294]
- Tile extent: 100

### Attribute ###
- Name: <anonymous>
- Type: FLOAT32
- Compressor: ZSTD
- Compression level: -1
- Cell val num: 1
```

**Comment by ihnorton** [comment]
@bkmartinjr I realize this may be hard to reproduce, so I made a set of test wheels with some additional debug output enabled in this error path.

- extra output in the error message above which should help us figure out why this is happening
- after (hopefully) getting the additional test output, please create a file called `/tmp/crashme` and run again with `catchseg python <yourcode.py>` -- this should cause a backtrace to be printed for the error path.

Thanks!

**Comment by bkmartinjr** [comment]
I found a test case.   The critical value is the size of the first dimension - making it "tile size + 1" triggers the bug.

When I run it, on Python 3.8, Ubuntu/WSL, tiledb 0.5.8:
```
$ python 230consolidate.py 
(30001, 33295)
rows 0 : 10000
rows 10000 : 20000
rows 20000 : 30000
rows 30000 : 30001
Traceback (most recent call last):
  File "230consolidate.py", line 38, in <module>
    main()
  File "230consolidate.py", line 32, in main
    tiledb.consolidate(X_name)
  File "tiledb/libtiledb.pyx", line 4678, in tiledb.libtiledb.consolidate
  File "tiledb/libtiledb.pyx", line 413, in tiledb.libtiledb._raise_ctx_err
  File "tiledb/libtiledb.pyx", line 398, in tiledb.libtiledb._raise_tiledb_error
tiledb.libtiledb.TileDBError: [TileDB::Query] Error: Subarray out of bounds. subarray: [0, 30049, 0, 33299] domain: [0, 30000, 0, 33294]
```

Code:
```
import tiledb
import numpy as np


def make_array(name, shape):

    filters = tiledb.FilterList([tiledb.ZstdFilter()])
    attrs = [
        tiledb.Attr(dtype=np.float32, filters=filters),
    ]
    domain = tiledb.Domain(tiledb.Dim(name="obs", domain=(0, shape[0] - 1), tile=min(shape[0], 50), dtype=np.uint32),
                           tiledb.Dim(name="var", domain=(0, shape[1] - 1), tile=min(shape[1], 100), dtype=np.uint32))
    schema = tiledb.ArraySchema(domain=domain, sparse=False, attrs=attrs,
                                cell_order='row-major', tile_order='col-major')
    tiledb.DenseArray.create(name, schema)


def main():
    X_name = "X"
    shape = (30001, 33295)
    print(shape)

    make_array(X_name, shape)

    stride = 10_000
    with tiledb.DenseArray(X_name, mode="w") as X:
        for row in range(0, shape[0], stride):
            lim = min(row + stride, shape[0])
            X[row:lim, :] = np.zeros((lim - row, shape[1]), dtype=np.float32)
            print("rows", row, ":", lim)

        tiledb.consolidate(X_name)

    tiledb.consolidate(X_name)


if __name__ == '__main__':
    main()
```

**Comment by ihnorton** [comment]
Hi @bkmartinjr, we will have wheels up with the fix today.

**Comment by ihnorton** [comment]
@bkmartinjr wheels for TileDB 0.5.9 with the fix are up on PyPI (https://pypi.org/project/tiledb/0.5.9/).

**Comment by bkmartinjr** [comment]
thanks - I am validating on the original dataset, and will close this if it is fixed.

**Comment by bkmartinjr** [comment]
It worked great.  Thanks for the resolution.
